Bug fix for VisItDataCollection name parsing

mlstowell opened the pull request:
When a `VisItDataCollection` name contains underscores it cannot be loaded by applications such as `miniapps/tools/get-values`. This is due to the fact that `VisItDataCollection::ParseVisItRootString` uses `string::find` rather than `string::rfind` to separate the `name` portion of the string from the cycle number.

This PR resolves the bug mentioned in issue #3328.
<!--GHEX{"author":"mlstowell","assignment":"2022-11-17T02:03:05","editor":"tzanio","id":3329,"reviewers":["termi-official","tzanio"],"merge":"2022-11-30T16:02:02","approval":"2022-11-28T22:32:42"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3329](https://github.com/mfem/mfem/pull/3329) | @mlstowell | @tzanio | @termi-official + @tzanio | 11/16/22 | 11/28/22 | 11/30/22 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

mlstowell commented:
@termi-official, you raise an interesting point. I didn't realize the pad digits was not working. I've added an option to `miniapps/tools/get-values` so that we can test this more easily in the future.

Using your modified `ex5` and `get-values` I was able to confirm that loading a `VisItDataCollection` seems to work correctly with arbitrary numbers of pad digits in the cycle number and MPI rank. Consequently, I believe the data collection is now working as expected. However, VisIt (version 2.13.3) could not access any data collections with the number of MPI rank pad digits different from 6. Various choices for the number of pad digits in the cycle all worked fine.

Based on these tests I believe VisIt (at least my version) does not support changing the number MPI rank pad digits. For this reason I have added a new override of `SetPadDigits` which only changes the cycle pad digits. I also added a note to the Doxygen comments explaining this.

mlstowell commented:
1. I'm not certain. We may ask @acfisher. I believe he wrote this initially. 

2. We make no assumptions regarding the time step. This is merely a piece of data that VisIt will display in some of its plots.

3. The cycle number is used by VisIt to order a series of datasets into a single grouping. Typically this is used for time series results but such a group could be used in a variety of ways e.g. a sequence of AMR refinements, Newton iterations, etc.. The cycle numbers are non-negative integers and they need not be sequential. VisIt will access them in numerical order. Repeated cycle numbers are allowed but the files will overwrite each other which may or may not be useful. 
> Since I read through related parts of the DataCollection logic, I have some comments which could be addressed in follow-up PRs, because the invariants of the data structure are not clear.
> 
> 1. Why is the collection name extracted in `ParseVisItRootString` from the mesh path and not in `LoadVisItRootFile` from the root file? If we have this rigid structure, then it does not make sense to store the full mesh path in the root file.
> 2. What is the role of `time_step`? I.e. do we assume that the solution if stored densely (each time step)?
> 3. What are the invariants regarding the cycle? I.e. do we have parts where we assume continuity of the indices? Do we have to pay attention when they are not continuous?

termi-official reviewed (changes requested):
Thanks for the elaboration. I think this fix is also valid. However, this patch leaves the remaining tools broken. This should work for now.

Is there anything blocking an update to C++17 as standard? This way we could pull in <filesystem> and find the padding. Alternatively there we should prove another file with such meta information (tried to address this in #2689, but its not a nice solution). Fixing this is probably out of scope for this PR...

acfisher commented:
Just to chime in here on my original implementation.  At the time I wrote it, Visit did not allow differences in the number of pad digits and I don't believe they changed that behavior.  So you should leave it at the default if you expect visit to be able to read the collections.  Also the format of the root file had constraints due to how visit uses them.  There is some redundancy there, but it's what it took to interoperate with VisIt

mlstowell commented on fem/datacollection.hpp:
```diff
@@ -457,6 +457,12 @@ class VisItDataCollection : public DataCollection
    virtual void RegisterQField(const std::string& q_field_name,
                                QuadratureFunction *qf);
 
+   /// Set the number of digits used for both the cycle and the MPI rank
+   /// @note VisIt seems to require 6 pad digits for the MPI rank. Therefore,
+   /// this function uses this default value. This behavior can be overridden
+   /// by calling SetPadDigitsCycle() and SetPadDigitsRank() instead.
+   void SetPadDigits(int digits) { pad_digits_cycle=digits; pad_digits_rank=6; }
+
```
Right, I forgot to change that.

mlstowell replied:
I thought it was best to leave `SetPadDigitsRank` as it is. If someone has a specific reason for altering this then we should let them. The current versions of VisIt may not be able to load the files but perhaps that is not their goal.

mlstowell replied:
I can alter `load-dc` and `convert-dc` to support different pad lengths. This should be a quick fix.

mlstowell replied:
Have you noticed any other programs that may be impacted?

termi-official replied:
As far as I am aware no other code uses the loading logic (minus the now failing test).

mlstowell replied:
Thanks @termi-official.

Marking this as `override` does sound correct and useful but there seems to be no precedent in MFEM. If we do this it would seem wiser to mark all of the overrides in this class at least (or even better all classes in this header file). Creating this type of precedent may require a vote from a wider group of developers. What do you say @tzanio?

mlstowell replied:
We discussed this at our weekly PR meeting and it seems I was mistaken, we do use `override` elsewhere in MFEM. I've added the `override` decoration to the relevant member functions of `VisItDataCollection`. Thanks for suggesting this, @termi-official.

termi-official replied:
Thanks for taking care!

termi-official commented:
We should keep in mind that a changelog entry is missing, because the change is breaking.

termi-official reviewed (approved):
LGTM.

A quick final, but optional suggestion. Maybe we should at least print some warning or comparable when `SetPadDigits` or `SetPadDigitsRank` is called to notify the user that something might go wrong down the line.

tzanio reviewed (approved):
Thanks @mlstowell !

tzanio commented:
Merged in `next` for testing...
